[FEA] dask-cudf doesn't support "corr"/correlation function like Pandas and cuDF

When attempting to perform a correlation like ```sales_corr = sales['pr_review_rating', 'count'].corr(sales['pr_review_rating', 'mean'])``` dask-cudf fails with the following error. 

```TypeError: cannot concatenate object of type "<class 'cudf.core.series.Series'>"; only pd.Series, pd.DataFrame, and pd.Panel (deprecated) objs are valid```

It seems that we might limit this currently. However dask-cudf should behave exactly like cuDF and Pandas. https://github.com/rapidsai/cudf/blob/4613ba821e4ed03a2db744f2c0bb0959fd450191/python/dask_cudf/dask_cudf/backends.py#L30-L33

Reproducer in a dev-nightly container as of 11/10/19:

```python
import dask_cudf
import cudf
​
df = cudf.datasets.randomdata(10)
ddf = dask_cudf.from_cudf(df, 2)
ddf.x.corr(ddf.y)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-3-4ce11b32272a> in <module>
      4 df = cudf.datasets.randomdata(10)
      5 ddf = dask_cudf.from_cudf(df, 2)
----> 6 ddf.x.corr(ddf.y)

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/dataframe/core.py in corr(self, other, method, min_periods, split_every)
   2978         if method != "pearson":
   2979             raise NotImplementedError("Only Pearson correlation has been implemented")
-> 2980         df = concat([self, other], axis=1)
   2981         return cov_corr(
   2982             df, min_periods, corr=True, scalar=True, split_every=split_every

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/dataframe/multi.py in concat(dfs, axis, join, interleave_partitions)
   1046     if axis == 1:
   1047         if all(df.known_divisions for df in dasks):
-> 1048             return concat_indexed_dataframes(dfs, axis=axis, join=join)
   1049         elif (
   1050             len(dasks) == len(dfs)

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/dataframe/multi.py in concat_indexed_dataframes(dfs, axis, join)
    884     warn = axis != 0
    885     meta = methods.concat(
--> 886         [df._meta for df in dfs], axis=axis, join=join, filter_warning=warn
    887     )
    888     empties = [strip_unknown_categories(df._meta) for df in dfs]

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask/dataframe/methods.py in concat(dfs, axis, join, uniform, filter_warning)
    354         func = concat_dispatch.dispatch(type(dfs[0]))
    355         return func(
--> 356             dfs, axis=axis, join=join, uniform=uniform, filter_warning=filter_warning
    357         )
    358 

/opt/conda/envs/rapids/lib/python3.7/site-packages/dask_cudf/backends.py in concat_cudf(dfs, axis, join, uniform, filter_warning, sort)
     31     dfs, axis=0, join="outer", uniform=False, filter_warning=True, sort=None
     32 ):
---> 33     assert axis == 0
     34     assert join == "outer"
     35     return cudf.concat(dfs)

AssertionError: 
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -31,6 +31,7 @@
 - PR #3291 Add normalize_nans_and_zeros
 - PR #3368 Enable dropna argument in dask_cudf groupby
 - PR #3298 add null replacement iterator for column_device_view
+- PR #3393 Implement df.cov and enable covariance/correlation in dask_cudf
 
 ## Improvements
 

--- a/python/cudf/cudf/core/dataframe.py
+++ b/python/cudf/cudf/core/dataframe.py
@@ -4167,6 +4167,23 @@ class DataFrame(object):
         else:
             return result
 
+    def cov(self, **kwargs):
+        """Compute the covariance matrix of a DataFrame.
+        Parameters
+        ----------
+        **kwargs
+            Keyword arguments to be passed to cupy.cov
+        Returns
+        -------
+        cov : DataFrame
+        """
+        cov = cupy.cov(self.values, rowvar=False)
+        df = DataFrame()
+        for i, col in enumerate(self.columns):
+            df[col] = cov[i]
+        df.index = self.columns
+        return df
+
 
 def from_pandas(obj):
     """

--- a/python/cudf/cudf/core/dataframe.py
+++ b/python/cudf/cudf/core/dataframe.py
@@ -4178,10 +4178,10 @@ class DataFrame(object):
         cov : DataFrame
         """
         cov = cupy.cov(self.values, rowvar=False)
-        df = DataFrame()
-        for i, col in enumerate(self.columns):
-            df[col] = cov[i]
-        df.index = self.columns
+        df = DataFrame.from_gpu_matrix(cupy.asfortranarray(cov)).set_index(
+            self.columns
+        )
+        df.columns = self.columns
         return df
 
 

--- a/python/cudf/cudf/tests/test_dataframe.py
+++ b/python/cudf/cudf/tests/test_dataframe.py
@@ -4054,3 +4054,22 @@ def test_insert(data):
     gdf.insert(-1, "qux", data)
 
     assert_eq(pdf, gdf)
+
+
+def test_cov():
+    gdf = gd.datasets.randomdata(10)
+    pdf = gdf.to_pandas()
+
+    assert_eq(pdf.cov(), gdf.cov())
+
+
+@pytest.mark.xfail(reason="cupy-based cov does not support nulls")
+def test_cov_nans():
+    pdf = pd.DataFrame()
+    pdf["a"] = [None, None, None, 2.00758632, None]
+    pdf["b"] = [0.36403686, None, None, None, None]
+    pdf["c"] = [None, None, None, 0.64882227, None]
+    pdf["d"] = [None, -1.46863125, None, 1.22477948, -0.06031689]
+    gdf = gd.from_pandas(pdf)
+
+    assert_eq(pdf.cov(), gdf.cov())

--- a/python/dask_cudf/dask_cudf/backends.py
+++ b/python/dask_cudf/dask_cudf/backends.py
@@ -30,9 +30,8 @@ def make_meta_cudf_index(x, index=None):
 def concat_cudf(
     dfs, axis=0, join="outer", uniform=False, filter_warning=True, sort=None
 ):
-    assert axis == 0
     assert join == "outer"
-    return cudf.concat(dfs)
+    return cudf.concat(dfs, axis=axis)
 
 
 try:

